### PR TITLE
Token based remote connection

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by turtle0x1** [comment]
Can tokens have some kind of `prefix` like `LT:` (LXD Token) or something like that so its easy to spot them in user input fields without custom requirements?

**Comment by stgraber** [comment]
We're not planning on doing that as we're modeling this after the existing cluster API.

On the client side, whether it's a trust password or a token not much happens other than if you want you can unpack it to get the address and fingerprint so you don't need to prompt for that.

The token itself being a valid password for the normal API was done so existing API clients can all use those unmodified and if they wish can get more information from them.

File: doc/authentication.md [hunk 1]
```diff
@@ -75,6 +76,13 @@ To allow establishing a new trust relationship from the client side, you must se
 In a production setup, unset `core.trust_password` after all clients have been added.
 This prevents brute-force attacks trying to guess the password.
 
+(authentication-token)=
+#### Adding client certificates using tokens
+
+New clients can be added using tokens. This is a safer way than using the trust password, as tokens expiry once they've been used. For this, the server needs to generate a token and pass it to the client.
+
+The server needs to call `lxc config trust add` which will prompt for a client name, and return a token. The client then calls `lxc remote add <name> <token>` to add the remote.
```
[the remark this answers is not in the document]
Comment (monstermunchkin) [reply doc/authentication.md:84]: It also works if you just run `lxc remote add <name>`. It will prompt for the trust password which also accepts a join token.

File: doc/authentication.md [hunk 2]
```diff
@@ -75,6 +76,13 @@ To allow establishing a new trust relationship from the client side, you must se
 In a production setup, unset `core.trust_password` after all clients have been added.
 This prevents brute-force attacks trying to guess the password.
 
+(authentication-token)=
+#### Adding client certificates using tokens
+
+New clients can be added using tokens. This is a safer way than using the trust password, as tokens expiry once they've been used. For this, the server needs to generate a token and pass it to the client.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/authentication.md:82]: Updated suggestion:
```suggestion
You can also add new clients by using tokens. This is a safer way than using the trust password, because tokens expire once they've been used. 

To use this method, generate a token for each client by calling `lxc config trust add`, which will prompt for the client name.
The clients can then add their certificates to the server's trust store by providing the generated token when prompted for the trust password. 
Alternatively, the clients can provide the token directly when adding the remote: `lxc remote add <name> <token>`
```

File: doc/authentication.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/authentication.md:84]: Great! That fits what we documented before, so no bigger changes needed. :)

We should clarify this above though - something like:
```
   - If the client certificate is not in the server's trust store, the server checks if a trust password is set or if there is a token for the client. If there is, the server prompts the user for it.
     If the provided token or trust password matches, the client certificate is added to the server's trust store and the connection is granted.
     Otherwise, the connection is rejected.
   - If the client certificate is not in the server's trust store and no trust password or token is set, the connection is rejected.
```
Actually, is that correct? Will the server prompt only if a trust password or token is set, or will it always prompt and reject if whatever is provided isn't correct?

[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/authentication.md:84]:
```suggestion
```
This line is covered in the suggestion for line 82, so this one should be removed.
Just let me know if you'd prefer me to add a commit with the doc changes instead of commenting. ;)

[the remark this answers is not in the document]
Comment (monstermunchkin) [reply doc/authentication.md:84]: > Actually, is that correct? Will the server prompt only if a trust password or token is set, or will it always prompt and reject if whatever is provided isn't correct?

It will always prompt. The client doesn't know if a password is set or if there's a valid pending token.

> Just let me know if you'd prefer me to add a commit with the doc changes instead of commenting. ;)

Commenting is perfectly fine ;)

[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/authentication.md:84]: OK, great, that makes the explanation a lot easier.
So lines 49-53 should be replaced with:
``` 
   - If the client certificate is in the server's trust store, the connection is granted.
   - If the client certificate is not in the server's trust store, the server prompts the user for a token or the trust password.
     If the provided token or trust password matches, the client certificate is added to the server's trust store and the connection is granted.
     Otherwise, the connection is rejected.

File: shared/api/certificate.go [hunk 3]
```diff
@@ -26,6 +26,12 @@ type CertificatesPost struct {
 	// Server trust password (used to add an untrusted client)
 	// Example: blah
 	Password string `json:"password" yaml:"password"`
+
+	// Whether to use token
```
Comment (stgraber) [inline shared/api/certificate.go:30]: That's incorrect, it doesn't intake that we're consuming a token, it indicates that we're issuing a token.

So `Whether to create a join token` would be better

File: shared/api/certificate.go [hunk 4]
```diff
@@ -72,3 +77,36 @@ type Certificate struct {
 func (cert *Certificate) Writable() CertificatePut {
 	return cert.CertificatePut
 }
+
+// ClientJoinToken represents the fields contained within an encoded client join token.
+//
+// swagger:model
+//
+// API extension: client_join_token
+type ClientJoinToken struct {
```
Comment (stgraber) [inline shared/api/certificate.go:92]: Probably should be prefixed with `Certificate` for consistency.

So maybe `CertificateAddToken`?

File: shared/api/certificate.go [hunk 5]
```diff
@@ -72,3 +77,36 @@ type Certificate struct {
 func (cert *Certificate) Writable() CertificatePut {
 	return cert.CertificatePut
 }
+
+// ClientJoinToken represents the fields contained within an encoded client join token.
+//
+// swagger:model
+//
+// API extension: client_join_token
```
Comment (stgraber) [inline shared/api/certificate.go:91]: API extension should be `certificate_token`

File: doc/api-extensions.md [hunk 6]
```diff
@@ -1564,11 +1564,11 @@ This adds `live-migrate` as a config option to `cluster.evacuate`, which forces
...
 
 ## instance\_allow\_inconsistent\_copy
-Adds `allow_inconsistent` field to instance source on `POST /1.0/instances`. If true, rsync will ignore the 
-`Partial transfer due to vanished source files` (code 24) error when creating an instance from a copy. 
+Adds `allow_inconsistent` field to instance source on `POST /1.0/instances`. If true, rsync will ignore the
+`Partial transfer due to vanished source files` (code 24) error when creating an instance from a copy.
 
 ## network\_state\_ovn
-This adds an "ovn" section to the /1.0/networks/NAME/state API which contains additional state information relevant to 
+This adds an "ovn" section to the /1.0/networks/NAME/state API which contains additional state information relevant to
```
Comment (stgraber) [inline doc/api-extensions.md:1571]: Can you split that to its own commit as it's an unrelated change?

File: doc/api-extensions.md [hunk 7]
```diff
@@ -1602,3 +1602,8 @@ Adds a new `GET /1.0/networks-acls/NAME/log` API to retrieve ACL firewall logs.
 
 ## storage\_zfs\_blocksize
 Introduces a new `zfs.blocksize` property for ZFS storage volumes which allows to set volume block size.
+
+## client\_join\_token
```
Comment (stgraber) [inline doc/api-extensions.md:1606]: certificate_token

File: shared/version/api.go [hunk 8]
```diff
@@ -314,6 +314,7 @@ var APIExtensions = []string{
 	"storage_zfs_reserve_space",
 	"network_acl_log",
 	"storage_zfs_blocksize",
+	"client_join_token",
```
Comment (stgraber) [inline shared/version/api.go:317]: certificate_token

File: doc/api-extensions.md [hunk 9]
```diff
@@ -1602,3 +1602,8 @@ Adds a new `GET /1.0/networks-acls/NAME/log` API to retrieve ACL firewall logs.
...
+This introduces client tokens which can be used for adding new remotes in a safer way.
```
Comment (stgraber) [inline doc/api-extensions.md:1607]: Should be `This adds token based certificate addition to the trust store as a safer alternative to a trust password`.

`remotes` are a concept specific to the Go client so probably best to avoid it be mentioned on the server side.

File: shared/cert.go [hunk 10]
```diff
@@ -425,6 +429,38 @@ func GetRemoteCertificate(address string, useragent string) (*x509.Certificate,
 	return resp.TLS.PeerCertificates[0], nil
 }
 
+// ClientJoinTokenDecode decodes a base64 and JSON encode join token.
+func ClientJoinTokenDecode(input string) (*api.ClientJoinToken, error) {
```
Comment (stgraber) [inline shared/cert.go:433]: CertificateTokenDecode I think

File: client/interfaces.go [hunk 11]
```diff
@@ -83,6 +83,7 @@ type InstanceServer interface {
 	CreateCertificate(certificate api.CertificatesPost) (err error)
 	UpdateCertificate(fingerprint string, certificate api.CertificatePut, ETag string) (err error)
 	DeleteCertificate(fingerprint string) (err error)
+	CreateClientJoinToken(certificate api.CertificatesPost) (op Operation, err error)
```
Comment (stgraber) [inline client/interfaces.go:86]: `CreateCertificateToken` and it should probably ensure that `certificate.Token` is set to True.

File: lxd/certificates.go [hunk 12]
```diff
@@ -500,10 +507,24 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
...
+				if err == nil {
+					// If so then check there is a matching join operation.
+					joinOp, err := clientJoinTokenValid(d, r, project.Default, joinToken)
+					if err != nil {
+						return response.InternalError(errors.Wrapf(err, "Failed during search for join token operation"))
+					}
+
+					if joinOp == nil {
+						return response.Forbidden(fmt.Errorf("No matching client join operation found"))
+					}
```
Comment (stgraber) [inline lxd/certificates.go:521]: I'm failing to see where we're pulling the `req` from the token creation here. It looks like as soon as we found a valid token, the user can feed us whatever they want through `req` and we'll just add it to the trust store.

This would be a huge security issue. It's very important that what's described in https://discuss.linuxcontainers.org/t/lxd-token-based-remote-connection/13114/2?u=stgraber is implemented.

**Comment by stgraber** [comment]
Unless I'm missing something, this implementation is missing pretty much all of the security relevant bits described in the spec, so this needs to be sorted and the test need to be made to confirm that such security issues don't exist.

There's also a bit of renaming that's needed to line things up better.

File: lxd/certificates.go (hunk 12 above)
Comment (monstermunchkin) [reply lxd/certificates.go:521]: > It looks like as soon as we found a valid token, the user can feed us whatever they want through `req` and we'll just add it to the trust store.

That would be possible. In order to prevent this, we could simply override `req`, can't we?

```go
req = api.CertificatesPost{
	CertificatePut: api.CertificatePut{
		Name: joinToken.ClientName,
		Type: "client",
	},
}
```

Comment (monstermunchkin) [reply lxd/certificates.go:521]: We'd also need to add `Restricted` and `Projects` as I missed those.

Comment (stgraber) [reply lxd/certificates.go:521]: Right, we need to completely discard the user provided req and make a new one from the operation metadata.

And indeed, restricted and projects are the more critical fields here as it's they get omitted or the user can override them, then they can escalate privileges and get control of the entire system.

**Comment by stgraber** [comment]
jenkins: test this please

File: doc/api-extensions.md [hunk 13]
```diff
@@ -1605,3 +1605,8 @@ Introduces a new `zfs.blocksize` property for ZFS storage volumes which allows t
 
 ## metrics\_cpu\_seconds
 This is used to detect whether LXD was fixed to output used CPU time in seconds rather than as milliseconds.
+
+## certificate\_token
+This adds token based certificate addition to the trust store as a safer alternative to a trust password.
```
Comment (ru-fu) [inline doc/api-extensions.md:1610]:
```suggestion
This adds token-based certificate addition to the trust store as a safer alternative to a trust password.
```

**Review by ru-fu — APPROVED** [review]
One more nitpick, but otherwise the docs look good now :)

File: shared/api/certificate.go [hunk 14]
```diff
@@ -72,3 +77,44 @@ type Certificate struct {
...
+
+	// The fingerprint of the network certificate
+	// Example: 57bb0ff4340b5bb28517e062023101adf788c37846dc8b619eb2c3cb4ef29436
+	Fingerprint string `json:"fingerprint" yaml:"fingerprint"`
+
+	// The addresses of the server
+	// Example: ["10.98.30.229:8443"]
+	Addresses []string `json:"addresses" yaml:"addresses"`
+
+	// The random join secret.
```
Comment (stgraber) [inline shared/api/certificate.go:105]: Our pattern for those doc strings is to not end with a period

File: shared/api/certificate.go [hunk 15]
```diff
@@ -72,3 +77,44 @@ type Certificate struct {
 func (cert *Certificate) Writable() CertificatePut {
 	return cert.CertificatePut
 }
+
+// CertificateAddToken represents the fields contained within an encoded client join token.
```
Comment (stgraber) [inline shared/api/certificate.go:87]: s/client join token/certificate add token/

File: shared/api/certificate.go [hunk 16]
```diff
@@ -72,3 +77,44 @@ type Certificate struct {
...
+	// List of allowed projects (applies when restricted)
+	// Example: ["default", "foo", "bar"]
+	Projects []string `json:"projects" yaml:"projects"`
+
+	// Whether to limit the certificate to listed projects
+	// Example: true
+	Restricted bool `json:"restricted" yaml:"restricted"`
+}
+
+// String encodes the client join token as JSON and then Base64.
```
Comment (stgraber) [inline shared/api/certificate.go:118]: s/client join token/certificate add token/

Comment (stgraber) [reply shared/api/certificate.go:118]: I think we normally spell it `base64`, no capital b

File: shared/api/certificate.go [hunk 17]
```diff
@@ -26,6 +26,12 @@ type CertificatesPost struct {
 	// Server trust password (used to add an untrusted client)
 	// Example: blah
 	Password string `json:"password" yaml:"password"`
+
+	// Whether to create a join token
```
Comment (stgraber) [inline shared/api/certificate.go:30]: certificate add token

File: doc/api-extensions.md.orig [hunk 18]
```diff
@@ -0,0 +1,1614 @@
+# API extensions
```
Comment (stgraber) [inline doc/api-extensions.md.orig:1]: Committed a `.orig` by accident here

File: shared/cert.go [hunk 19]
```diff
@@ -425,6 +429,38 @@ func GetRemoteCertificate(address string, useragent string) (*x509.Certificate,
...
+		return nil, fmt.Errorf("No server addresses in join token")
+	}
+
+	if j.Secret == "" {
+		return nil, fmt.Errorf("No secret in join token")
+	}
+
+	if j.Fingerprint == "" {
+		return nil, fmt.Errorf("No certificate fingerprint in join token")
+	}
```
Comment (stgraber) [inline shared/cert.go:459]: certificate add token

File: lxd/certificates.go [hunk 20]
```diff
@@ -354,6 +354,88 @@ func clusterMemberJoinTokenValid(d *Daemon, r *http.Request, projectName string,
 	return nil, nil
 }
 
+// certificateTokenValid searches for certificate token that matches the joint token provided.
```
Comment (stgraber) [inline lxd/certificates.go:359]: `add token provided`

File: lxd/certificates.go [hunk 21]
```diff
@@ -534,10 +545,33 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
...
+					joinOp, err := certificateTokenValid(d, r, project.Default, joinToken)
+					if err != nil {
+						return response.InternalError(errors.Wrapf(err, "Failed during search for join token operation"))
+					}
+
+					if joinOp == nil {
+						return response.Forbidden(fmt.Errorf("No matching client join operation found"))
+					}
+
+					req = api.CertificatesPost{
```
Comment (stgraber) [inline lxd/certificates.go:561]: Worth adding a comment with something like:
```
// Create a new request from the token data as the user isn't allowed to override anything.
```

File: lxd/certificates.go [hunk 22]
```diff
@@ -354,6 +354,88 @@ func clusterMemberJoinTokenValid(d *Daemon, r *http.Request, projectName string,
...
+			continue
+		}
+
+		opProjects, ok := op.Metadata["projects"]
+		if !ok {
+			continue
+		}
+
+		projects := opProjects.([]string)
+
```
Comment (stgraber) [inline lxd/certificates.go:403]: can drop that line break

File: shared/api/certificate.go [hunk 23]
```diff
@@ -72,3 +77,44 @@ type Certificate struct {
...
+	// Example: 2b2284d44db32675923fe0d2020477e0e9be11801ff70c435e032b97028c35cd
+	Secret string `json:"secret" yaml:"secret"`
+
+	// List of allowed projects (applies when restricted)
+	// Example: ["default", "foo", "bar"]
+	Projects []string `json:"projects" yaml:"projects"`
+
+	// Whether to limit the certificate to listed projects
+	// Example: true
+	Restricted bool `json:"restricted" yaml:"restricted"`
```
Comment (stgraber) [inline shared/api/certificate.go:115]: This should not be in the add token.

The add token should only include what the client needs. The list of projects and whether it's restricted is something the administrator has set and that should remain on the server side.

File: lxd/certificates.go [hunk 24]
```diff
@@ -354,6 +354,88 @@ func clusterMemberJoinTokenValid(d *Daemon, r *http.Request, projectName string,
 	return nil, nil
 }
 
+// certificateTokenValid searches for certificate token that matches the joint token provided.
+// Returns matching operation if found and cancels the operation, otherwise returns nil.
+func certificateTokenValid(d *Daemon, r *http.Request, projectName string, joinToken *api.CertificateAddToken) (*api.Operation, error) {
```
Comment (stgraber) [inline lxd/certificates.go:361]: So this feels wrong.

What I'm expecting is:
 - Admin issues a token through `POST /1.0/certificates`
 - The `req` (CertificatesPost) data is stored in a metadata field of a token operation
 - The admin is returned the token which includes the address of the server, fingerprint and secret

 - The client uses the token, connects to the server at the address provided, validates the fingerprint using the one provided, then adds itself to the trust store by doing `POST /1.0/certificates` with the token
 - The server locates the operation for that token, retrieves the initial `req` from the operation and adds the client's certificate to the trust store using that recorded `req`

That means the user when provided a token will not necessarily know what kind of restrictions the administrator may have put in place on it and that's kinda the point as providing that information may cause a data leak of some kind.

File: lxd/certificates.go [hunk 25]
```diff
@@ -354,6 +354,50 @@ func clusterMemberJoinTokenValid(d *Daemon, r *http.Request, projectName string,
 	return nil, nil
 }
 
+// certificateTokenValid searches for certificate token that matches the add token provided.
+// Returns matching operation if found and cancels the operation, otherwise returns nil.
+func certificateTokenValid(d *Daemon, r *http.Request, projectName string, addToken *api.CertificateAddToken) (*api.Operation, error) {
+	ops, err := operationsGetByType(d, r, projectName, db.OperationCertificateAddToken)
```
Comment (stgraber) [inline lxd/certificates.go:362]: `projectName` shouldn't really be relevant here as certificates are considered a global resource so will always be in default

File: lxd/certificates.go [hunk 26]
```diff
@@ -354,6 +354,50 @@ func clusterMemberJoinTokenValid(d *Daemon, r *http.Request, projectName string,
...
+		opSecret, ok := op.Metadata["secret"]
+		if !ok {
+			continue
+		}
+
+		opFingerprint, ok := op.Metadata["fingerprint"]
+		if !ok {
+			continue
+		}
+
```
Comment (stgraber) [inline lxd/certificates.go:382]: No need to validate the fingerprint, it's a field that's useful to the client to connect but if the server gets its certificate replaced it shouldn't make all tokens invalid.

File: lxd/certificates.go [hunk 27]
```diff
@@ -496,10 +507,39 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
...
+						return response.InternalError(errors.Wrapf(err, "Failed during search for certificate add token operation"))
+					}
+
+					if joinOp == nil {
+						return response.Forbidden(fmt.Errorf("No matching certificate add operation found"))
+					}
+
+					tokenReq, ok := joinOp.Metadata["request"].(api.CertificatesPost)
+					if !ok {
+						return response.InternalError(fmt.Errorf("Failed during converting request"))
```
Comment (stgraber) [inline lxd/certificates.go:525]: That error looks confusing, how about `Bad certificate add operation data`?

File: lxd/certificates.go [hunk 28]
```diff
@@ -543,6 +583,48 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
...
 		name = remoteHost
+	} else if req.Token {
+		address, err := node.HTTPSAddress(d.db)
+		if err != nil {
+			return response.InternalError(err)
+		}
+		addresses, err := util.ListenAddresses(address)
+		if err != nil {
+			return response.InternalError(err)
+		}
```
Comment (stgraber) [inline lxd/certificates.go:598]: Could do with a comment

File: lxd/certificates.go [hunk 29]
```diff
@@ -543,6 +583,48 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
...
+
+		if req.Projects == nil {
+			req.Projects = []string{}
+		}
+
+		meta := map[string]interface{}{
+			"secret":      joinSecret,
+			"fingerprint": fingerprint,
+			"addresses":   addresses,
```
Comment (stgraber) [inline lxd/certificates.go:622]: No need to put those two in the operation data

File: lxc/config_trust.go [hunk 30]
```diff
@@ -123,34 +123,49 @@ func (c *cmdConfigTrustAdd) Run(cmd *cobra.Command, args []string) error {
...
+	cert := api.CertificatesPost{}
 
-	var name string
-	if c.flagName != "" {
-		name = c.flagName
+	if len(args) == 0 {
+		// Use token
+		cert.Token = true
+
+		cert.Name, err = cli.AskString("Please provide client name: ", "", nil)
```
Comment (stgraber) [inline lxc/config_trust.go:140]: Needs to be wrapped with `i18n.G`

File: lxc/config_trust.go [hunk 31]
```diff
@@ -123,34 +123,49 @@ func (c *cmdConfigTrustAdd) Run(cmd *cobra.Command, args []string) error {
...
+
+		cert.Certificate = base64.StdEncoding.EncodeToString(x509Cert.Raw)
+		cert.Name = name
+	}
 
 	if c.flagType == "client" {
 		cert.Type = api.CertificateTypeClient
 	} else if c.flagType == "metrics" {
+		if cert.Token {
+			return errors.New("Cannot use metrics type certificate when using a token")
```
Comment (stgraber) [inline lxc/config_trust.go:174]: We use `fmt.Errorf` and error text should be wrapped in `i18n.G`

File: lxc/config_trust.go [hunk 32]
```diff
@@ -159,9 +174,73 @@ func (c *cmdConfigTrustAdd) Run(cmd *cobra.Command, args []string) error {
...
+		op, err := resource.server.CreateCertificateToken(cert)
+		if err != nil {
+			return err
+		}
+
+		if !c.global.flagQuiet {
+			opAPI := op.Get()
+			joinToken, err := certificateAddTokenOperationToAPI(&opAPI)
+			if err != nil {
+				return fmt.Errorf("Failed converting token operation to join token: %w", err)
```
Comment (stgraber) [inline lxc/config_trust.go:195]: i18n.G

File: lxc/config_trust.go [hunk 33]
```diff
@@ -159,9 +174,73 @@ func (c *cmdConfigTrustAdd) Run(cmd *cobra.Command, args []string) error {
...
+			fmt.Printf(i18n.G("Client %s join token: %s")+"\n", cert.Name, joinToken.String())
+		}
+
+		return nil
+	}
+
 	return resource.server.CreateCertificate(cert)
 }
 
+func certificateAddTokenOperationToAPI(op *api.Operation) (*api.CertificateAddToken, error) {
```
Comment (stgraber) [inline lxc/config_trust.go:207]: Maybe make this a function on an `api.Operation`?

Like `op.ToCertificateAddToken(*CertificateAddToken, error)`?

File: lxc/remote.go [hunk 34]
```diff
@@ -83,7 +83,7 @@ type cmdRemoteAdd struct {
 
 func (c *cmdRemoteAdd) Command() *cobra.Command {
 	cmd := &cobra.Command{}
-	cmd.Use = usage("add", i18n.G("[<remote>] <IP|FQDN|URL>"))
+	cmd.Use = usage("add", i18n.G("[<remote>] <IP|FQDN|URL|Token>"))
```
Comment (stgraber) [inline lxc/remote.go:86]: I'd write it as `token`

File: lxc/remote.go [hunk 35]
```diff
@@ -150,6 +150,148 @@ func (c *cmdRemoteAdd) findProject(d lxd.InstanceServer, project string) (string
 	return project, nil
 }
 
+func (c *cmdRemoteAdd) parseURL(addr string) (string, error) {
```
Comment (stgraber) [inline lxc/remote.go:153]: This looks way overly complex.

The addresses in a join token can only ever be https, so we don't really need all of the checks around scheme and unix and whatnot here.

You can pretty much just prefix the address with `https://` and call it done.

File: lxc/config_trust.go [hunk 36]
```diff
@@ -572,6 +576,77 @@ func (c *cmdConfigTrustRemove) Run(cmd *cobra.Command, args []string) error {
...
+
+		if joinToken.ClientName == resource.name {
+			// Delete the operation
+			err = resource.server.DeleteOperation(op.ID)
+			if err != nil {
+				return err
+			}
+
+			if !c.global.flagQuiet {
+				fmt.Printf(i18n.G("certificate add token for %s deleted")+"\n", resource.name)
```
Comment (stgraber) [inline lxc/config_trust.go:640]: `Certificate`

File: lxc/cluster.go [hunk 37]
```diff
@@ -765,7 +768,14 @@ func (c *cmdClusterAdd) Run(cmd *cobra.Command, args []string) error {
 	resource := resources[0]
 
 	if resource.name == "" {
-		return fmt.Errorf(i18n.G("Missing cluster member name"))
+		if c.flagName == "" {
+			resource.name, err = cli.AskString("Please provide cluster member name: ", "", nil)
```
Comment (stgraber) [inline lxc/cluster.go:772]: `i18n.G`

File: lxc/config_trust.go (hunk 33 above)
Comment (monstermunchkin) [reply lxc/config_trust.go:207]: I don't mind. Should we do this for the cluster join tokens as well then? Currently that uses `clusterJoinTokenOperationToAPI(op *api.Operation) (*api.ClusterMemberJoinToken, error)`.

File: lxd/certificates.go (hunk 29 above)
Comment (monstermunchkin) [reply lxd/certificates.go:622]: These fields are in the operation data because the admin uses this data to construct the token. If these were missing, the admin would need to call `GetServer()` to get the addresses and fingerprint which IMHO is an unnecessary call.

File: lxc/config_trust.go (hunk 33 above)
Comment (stgraber) [reply lxc/config_trust.go]: Yep, that'd be great

File: lxd/certificates.go (hunk 29 above)
Comment (stgraber) [reply lxd/certificates.go:608]: Odd but nevermind, let's keep it the same as cluster tokens then :)

File: lxc/config_trust.go (hunk 33 above)
Comment (monstermunchkin) [reply lxc/config_trust.go]: Done
